### ISSUE TITLE
NaN values doing fine tuning with gpt-oss:20b

### ISSUE THREAD

**Issue by nfsecurity** [ml-explore/mlx-lm#361]
First of all, many thanks to all the people who collaborate with the mlx project.

I am trying to fine tune gpt-oss:20b but I am receiving "nan" values in the Train Loss and Value Loss in every iteration. 

I am using the latest mlx-lm in an Mac Studio M1 Ultra 64GB:

```
% pip3.11 show mlx-lm

Name: mlx-lm
Version: 0.26.3
Summary: LLMs with MLX and the Hugging Face Hub
Home-page: https://github.com/ml-explore/mlx-lm
Author: MLX Contributors
Author-email: mlx@group.apple.com
License: MIT
Location: /opt/homebrew/lib/python3.11/site-packages
Requires: jinja2, mlx, numpy, protobuf, pyyaml, transformers
Required-by: 

% pip3.11 show mlx   

Name: mlx
Version: 0.27.1
Summary: A framework for machine learning on Apple silicon.
Home-page: https://github.com/ml-explore/mlx
Author: MLX Contributors
Author-email: mlx@group.apple.com
License: MIT
Location: /opt/homebrew/lib/python3.11/site-packages
Requires: mlx-metal
Required-by: mlx-lm
```

Here are the steps I am following:

**Download the model**

```
% mlx_lm.convert --hf-path openai/gpt-oss-20b
```
I also tried with "--dtype bfloat16" and "-q --q-bits 8", but the result is the same.

**Create config.yaml file**
```
model: "./mlx_model"
train: true
fine_tune_type: lora
data: "dataset"
seed: 0
num_layers: 16
batch_size: 1

iters: 50
val_batches: 25
learning_rate: 1e-5
steps_per_report: 10
steps_per_eval: 100
resume_adapter_file: null
adapter_path: "adapters"
save_every: 100
test: true
test_batches: -1
max_seq_length: 8192
grad_checkpoint: true
lora_parameters:
  keys: [
    "self_attn.q_proj",
    "self_attn.v_proj",
    "self_attn.k_proj",
    "self_attn.o_proj",
    "mlp.gate_proj",
    "mlp.down_proj",
    "mlp.up_proj"
  ]
  rank: 8
  scale: 2.0
  dropout: 0.0

lr_schedule: false
```
I also tried deleting "keys" parameters, but the result is the same.

**Run training**
```
% mlx_lm.lora --train -c config.yaml
Loading configuration file config.yaml
Loading pretrained model
Loading datasets
Training
Trainable parameters: 0.013% (2.654M/20914.756M)
Starting training..., iters: 50
Calculating loss...: 100%|██████████████████████| 25/25 [00:15<00:00,  1.62it/s]
Iter 1: Val loss 4.951, Val took 15.467s
Iter 10: Train loss nan, Learning Rate 1.000e-05, It/sec 0.670, Tokens/sec 441.976, Trained Tokens 6598, Peak mem 46.413 GB
Iter 20: Train loss nan, Learning Rate 1.000e-05, It/sec 0.746, Tokens/sec 454.473, Trained Tokens 12691, Peak mem 46.413 GB
Iter 30: Train loss nan, Learning Rate 1.000e-05, It/sec 0.748, Tokens/sec 450.543, Trained Tokens 18714, Peak mem 46.413 GB
Iter 40: Train loss nan, Learning Rate 1.000e-05, It/sec 0.756, Tokens/sec 445.991, Trained Tokens 24611, Peak mem 46.413 GB
Calculating loss...: 100%|██████████████████████| 25/25 [00:14<00:00,  1.78it/s]
Iter 50: Val loss nan, Val took 14.056s
Iter 50: Train loss nan, Learning Rate 1.000e-05, It/sec 0.664, Tokens/sec 454.652, Trained Tokens 31459, Peak mem 46.413 GB
Saved final weights to adapters/adapters.safetensors.
Testing
Calculating loss...: 300it [03:00,  1.67it/s]
Test loss nan, Test ppl nan.
```

My JSONL files (train, validation and test) have the correct format required for the model (I think):
```
{
 "messages": [
   {
     "role":"user",
     "content":"user question"
   },
   {
    "role":"assistant",
    "content":"model answer"
   }
 ]
}
```
And of course, because they are JSONL files, a real one looks like this:
```
{"messages": [{"role":"user","content":"user question 1"},{"role":"assistant","content":"model answer 1"}]}
{"messages": [{"role":"user","content":"user question 2"},{"role":"assistant","content":"model answer 2"}]}
{"messages": [{"role":"user","content":"user question 3"},{"role":"assistant","content":"model answer 3"}]}
{"messages": [{"role":"user","content":"user question 4"},{"role":"assistant","content":"model answer 4"}]}
```
I am not using the "reasoning part" yet because I am first testing if the fine tuning works or not.

What am I doing wrong?

**Comment by nfsecurity** [comment]
I tried changing my JSONL format, this time including the "thinking part":

```
{
 "messages": [
   {
     "role":"user",
     "content":"user question"
   },
   {
    "role":"assistant",
    "content":"lets think about ..."
   },
   {
    "role":"assistant",
    "content":"model answer"
   }
 ]
}
```

And of course, because they are JSONL files, a real one looks like this:

```
{"messages": [{"role":"user","content":"user question 1"},{"role":"assistant","content":"lets think about ..."},{"role":"assistant","content":"model answer 1"}]}
{"messages": [{"role":"user","content":"user question 1"},{"role":"assistant","content":"lets think about ..."},{"role":"assistant","content":"model answer 1"}]}
{"messages": [{"role":"user","content":"user question 1"},{"role":"assistant","content":"lets think about ..."},{"role":"assistant","content":"model answer 1"}]}
{"messages": [{"role":"user","content":"user question 1"},{"role":"assistant","content":"lets think about ..."},{"role":"assistant","content":"model answer 1"}]}
```

But the result at training time is the same (NAN values):

```
% mlx_lm.lora --train -c config.yaml
Loading configuration file config.yaml
Loading pretrained model
Loading datasets
Training
Trainable parameters: 0.013% (2.654M/20914.756M)
Starting training..., iters: 50
Calculating loss...: 100%|███████████████████████| 25/25 [00:30<00:00,  1.22s/it]
Iter 1: Val loss 3.806, Val took 30.572s
Iter 10: Train loss nan, Learning Rate 1.000e-05, It/sec 0.360, Tokens/sec 457.432, Trained Tokens 12707, Peak mem 48.869 GB
Iter 20: Train loss nan, Learning Rate 1.000e-05, It/sec 0.368, Tokens/sec 459.675, Trained Tokens 25203, Peak mem 49.915 GB
Iter 30: Train loss nan, Learning Rate 1.000e-05, It/sec 0.355, Tokens/sec 461.215, Trained Tokens 38205, Peak mem 50.187 GB
Iter 40: Train loss nan, Learning Rate 1.000e-05, It/sec 0.346, Tokens/sec 452.956, Trained Tokens 51296, Peak mem 50.390 GB
Calculating loss...: 100%|███████████████████████| 25/25 [00:28<00:00,  1.14s/it]
Iter 50: Val loss nan, Val took 28.577s
Iter 50: Train loss nan, Learning Rate 1.000e-05, It/sec 0.343, Tokens/sec 456.318, Trained Tokens 64592, Peak mem 50.952 GB
Saved final weights to adapters/adapters.safetensors.
Testing
Calculating loss...: 300it [05:48,  1.16s/it]
Test loss nan, Test ppl nan.
```

This time the training starts with less Val loss (3.806 instead of 4.951, maybe because of the thinking part helps a lot) but the NAN values are still present in each iteration.

**Comment by nfsecurity** [comment]
I am digging more on this. I tried to fine tune gpt-oss because I read this on the last MLX-LM update:

_**Add Additional Features of GPT-OSS Model : Lora, Alternating attention, MoE Support by @Shashikant86 in https://github.com/ml-explore/mlx-lm/pull/357**_

But now, I am very curious about the MLX "technique" for fine tuning GPT-OSS. The Unsloth team release yesterday [this blog](https://unsloth.ai/blog/gpt-oss), and they said:

"_We found that the gpt-oss' MXFP4 format while being efficient, doesn't support training. So, we had to implement custom training functions for the layers that use MXFP4 to make it work while ensuring we do not compromise on speed. We utilized OpenAI's Triton Kernels library directly to allow MXFP4 inference. For finetuning / training however, the MXFP4 kernels do not yet support training, since the backwards pass is not yet implemented. We're actively working on implementing it in Triton! There is a flag called W_TRANSPOSE as mentioned , which should be implemented. The derivative can be calculated by the transpose of the weight matrices, and so we have to implement the transpose operation. As a result, if you want to train gpt-oss with any library other than Unsloth, you’ll need to upcast the weights to bf16 before training. This approach, however, significantly increases both VRAM usage and training time by as much as 300% more memory usage! ALL other training methods will require a minimum of 65GB VRAM to train the 20b model while Unsloth only requires 14GB VRAM (-80%). Explanation: Both models are Mixture of Experts (MoE), and the smaller 20B model choses top 4 experts out of 32 total experts, its bigger brother, the 120B activates 4 experts per token out of the total 128 experts. During training (and release), the MoE is set to use MXFP4 format for its weights. On top of that, the weights are stored as Parameter (nn.Paramter) instead of being stored as linear layers (nn.Linear) like it is generally done. This makes quantisation hard. Especially since this is MoE and MLP/Experts contributes to approximately 19B of the total 20B parameters. So quantising this and performing the operations efficiently becomes crucial. To make quantization with BitsandBytes, we converted the weights and the architecture from the aforementioned Parameter format to the Linear format. Thus, gpt-oss-20b finetuning is able to fit in under 14GB of VRAM with Unsloth!_"

That sounds very complicated. My question here: Is MLX-LM ready now to fine tune GPT-OSS ? or Am I trying to fine tune an unsupported model (for now) ?

Thank you again for your amazing work.

**Comment by ChristopherLyon** [comment]
Bumping this - would love to tune gpt-oss on my M3

**Comment by nfsecurity** [comment]
A little update: I made a mistake in the final JSONL format (sorry), the correct format is:

```
{"text":"<|start|>system<|message|>You are an expert in ...<|end|><|start|>user<|message|>user question 1 ...<|end|><|start|>assistant<|channel|>analysis<|message|>Lets think about this ...<|end|><|start|>assistant<|channel|>final<|message|>Final answer ...<|return|>"}
{"text":"<|start|>system<|message|>You are an expert in ...<|end|><|start|>user<|message|>user question 2 ...<|end|><|start|>assistant<|channel|>analysis<|message|>Lets think about this ...<|end|><|start|>assistant<|channel|>final<|message|>Final answer ...<|return|>"}
{"text":"<|start|>system<|message|>You are an expert in ...<|end|><|start|>user<|message|>user question 3 ...<|end|><|start|>assistant<|channel|>analysis<|message|>Lets think about this ...<|end|><|start|>assistant<|channel|>final<|message|>Final answer ...<|return|>"}
```

I am still getting NaN values:

```
% mlx_lm.lora --train -c config.yaml
Loading configuration file config.yaml
Loading pretrained model
Loading datasets
Training
Trainable parameters: 0.013% (2.654M/20914.756M)
Starting training..., iters: 50
Calculating loss...: 100%|██████████████████████| 2/2 [00:01<00:00,  1.95it/s]
Iter 1: Val loss 4.359, Val took 1.048s
Iter 10: Train loss nan, Learning Rate 1.000e-05, It/sec 1.272, Tokens/sec 560.084, Trained Tokens 4404, Peak mem 46.001 GB
Iter 20: Train loss nan, Learning Rate 1.000e-05, It/sec 1.310, Tokens/sec 576.863, Trained Tokens 8808, Peak mem 46.001 GB
Iter 30: Train loss nan, Learning Rate 1.000e-05, It/sec 1.306, Tokens/sec 575.018, Trained Tokens 13212, Peak mem 46.001 GB
Iter 40: Train loss nan, Learning Rate 1.000e-05, It/sec 1.309, Tokens/sec 576.526, Trained Tokens 17616, Peak mem 46.001 GB
Calculating loss...: 100%|██████████████████████| 2/2 [00:00<00:00,  2.32it/s]
Iter 50: Val loss nan, Val took 0.865s
Iter 50: Train loss nan, Learning Rate 1.000e-05, It/sec 1.307, Tokens/sec 575.396, Trained Tokens 22020, Peak mem 46.001 GB
Saved final weights to adapters/adapters.safetensors.
Testing
Calculating loss...: 2it [00:00,  2.32it/s]
Test loss nan, Test ppl nan.
```

**Comment by ethicalvats** [comment]
looking into this too...

I tried to run finetuning with this branch of Unsloth https://github.com/unslothai/unsloth/pull/1289/files

However, later inspecting the code I found that it is missing implementation for gpt_oss inside mlx package.

@Totorong-Bot where did you found the mlx implementation of gpt_oss?

**Comment by gregtozzi** [comment]
Just had the same problem with a Qwen fine tune. Try setting your learning-rate to a float. So 1.0e-5 rather than 1e-5.
